app/services/katello/pulp3/content_view_version/import_validator.rb

```diff
@@ -26,113 +27,107 @@ def check!
-          bad_repos = interested_repos.select do |repo|
-            repo_in_metadata = metadata_map[[repo.product.label, repo.label]]
-            repo.redhat? != repo_in_metadata[:redhat] ||
-              repo.slice(:name, :label, :content_type) != repo_in_metadata.slice(:name, :label, :content_type)
+            !(katello_repo.content_type == metadata_repo.content_type &&
+              katello_repo.redhat? == metadata_repo.redhat)
           end
 
           if bad_repos.any?
-            repos_to_report = bad_repos.map { |repo| [repo.product.label, repo.label] }
             fail _("The following repositories provided in the import metadata have an incorrect content type or provider type. "\
                     "Make sure the export and import repositories are of the same type before importing\n "\
-                    "%{repos}" % { content_view: content_view.name,
-                                   repos: generate_product_repo_i18n_string(repos_to_report).join("")}
+                    "%{repos}" % { content_view: @content_view.name,
+                                   repos: generate_product_repo_i18n_string(bad_repos).join("")}
                   )
           end
         end
 
         def ensure_redhat_products_metadata_are_in_the_library!
-          products_in_library = ::Katello::Product.in_org(organization).redhat.pluck(:label)
-
-          # Map products to repositories in the metadata
-          # {product_label: [repositories with this product]}
-          products_in_metadata_map = {}
+          missing = @metadata_map.repositories.select do |repo|
+            repo.redhat && katello_product_for_metadata_repo(repo).nil?
+          end
 
-          metadata[:repositories].values.each do |repo|
-            next unless repo[:redhat]
-            products_in_metadata_map[repo[:product][:label]] ||= []
-            products_in_metadata_map[repo[:product][:label]] << repo
+          if missing.any?
+            repos_in_import = generate_product_repo_i18n_string(missing)
+            fail _("The organization's manifest does not contain the subscriptions required to enable the following repositories.\n "\
+                    "%{repos}" % { content_view: @content_view.name, repos: repos_in_import.join("")}
+                  )
           end
+        end
+
+        private
 
-          products_not_in_library = products_in_metadata_map.keys - products_in_library
-
-          unless products_not_in_library.blank?
-            # make a list of [product_name, repo_name] pairs to report to the user
-            product_repos = []
-            products_not_in_library.each do |prod|
-              products_in_metadata_map[prod].each do |repo|
-                product_name = metadata[:products][prod][:name]
-                repo_name = repo[:name]
-                product_repos << [product_name, repo_name]
-              end
+        def katello_product_for_metadata_repo(metadata_repo)
```

Comment:
I wasn't able to fix the original error by patching in this PR. It looks like it's due to my manifest.json not having a `cp_id` for the product.  That causes this method to search by label, which fails.  Why would the manifest.json file be missing the `cp_id`?

Reply:
I did a complete library export:

```
hammer content-export complete library --organization "export org"
```

and then a complete library import:

```
hammer content-import library --organization "import org" --metadata-file metadata-5.json  --path '/var/lib/pulp/exports/export_org/Export-Library/5.0/2022-03-08T21-11-36-00-00'
```

Reply:
My metadata file:

```
[vagrant@centos7-hammer-devel ~]$ cat metadata-5.json | jq
{
  "organization": "export org",
  "repositories": {
    "Red_Hat_Satellite_Tools_6_10_for_RHEL_8_x86_64_RPMs-1746498": {
      "name": "Red Hat Satellite Tools 6.10 for RHEL 8 x86_64 RPMs",
      "label": "Red_Hat_Satellite_Tools_6_10_for_RHEL_8_x86_64_RPMs",
      "description": null,
      "arch": "noarch",
      "content_type": "yum",
      "unprotected": false,
      "checksum_type": null,
      "os_versions": null,
      "major": null,
      "minor": null,
      "download_policy": "immediate",
      "mirroring_policy": "mirror_complete",
      "product": {
        "label": "Red_Hat_Enterprise_Linux_for_x86_64"
      },
      "gpg_key": {},
      "content": {
        "id": "12092",
        "label": "satellite-tools-6.10-for-rhel-8-x86_64-rpms"
      },
      "redhat": true
    }
  },
  "content_view": {
    "name": "Export-Library",
    "label": "Export-Library",
    "description": null,
    "generated_for": "library_export"
  },
  "content_view_version": {
    "major": 5,
    "minor": 0,
    "description": ""
  },
  "incremental": false,
  "products": {
    "Red_Hat_Enterprise_Linux_for_x86_64": {
      "name": "Red Hat Enterprise Linux for x86_64",
      "label": "Red_Hat_Enterprise_Linux_for_x86_64",
      "description": null,
      "gpg_key": {},
      "redhat": true
    }
  },
  "gpg_keys": {},
  "toc": "export-e6fca860-ad70-46d1-bed9-80f434abe6d4-20220308_2111-toc.json"
}
```

Reply:
Huh. I exported again with your PR in place and now I see the `cp_id`:

```
[vagrant@centos7-hammer-devel ~]$ cat metadata-6.json | jq
{
  "organization": "export org",
  "repositories": {
    "Red_Hat_Satellite_Tools_6_10_for_RHEL_8_x86_64_RPMs-1746498": {
      "name": "Red Hat Satellite Tools 6.10 for RHEL 8 x86_64 RPMs",
      "label": "Red_Hat_Satellite_Tools_6_10_for_RHEL_8_x86_64_RPMs",
      "description": null,
      "arch": "noarch",
      "content_type": "yum",
      "unprotected": false,
      "checksum_type": null,
      "os_versions": null,
      "major": null,
      "minor": null,
      "download_policy": "immediate",
      "mirroring_policy": "mirror_complete",
      "product": {
        "label": "Red_Hat_Enterprise_Linux_for_x86_64"
      },
      "gpg_key": {},
      "content": {
        "id": "12092",
        "label": "satellite-tools-6.10-for-rhel-8-x86_64-rpms"
      },
      "redhat": true
    }
  },
  "content_view": {
    "name": "Export-Library",
    "label": "Export-Library",
    "description": null,
    "generated_for": "library_export"
  },
  "content_view_version": {
    "major": 6,
    "minor": 0,
    "description": ""
  },
  "incremental": false,
  "products": {
    "Red_Hat_Enterprise_Linux_for_x86_64": {
      "name": "Red Hat Enterprise Linux for x86_64",
      "label": "Red_Hat_Enterprise_Linux_for_x86_64",
      "description": null,
      "cp_id": "479",
      "gpg_key": {},
      "redhat": true
    }
  },
  "gpg_keys": {},
  "toc": "export-275d3181-d043-4ee0-9f1e-cd6e45d8ce61-20220308_2151-toc.json"
}
```

Reply:
Oh I see, you added the `cp_id`!  Okay, so I guess then in the reproducer steps there should be another export.

Reply:
Sorry, I've been looking @ this pr a bit too long. You would need to re-export just as you found to get the cp_id in the metadata.json !

Reply:
Your test was actually a good one though, to test that error would still show up in absence of cp_id + mismatched repo labels